lib/core/string.toit

```diff
@@ -244,6 +244,65 @@ abstract class string implements Comparable:
       rune := this[i]
       if rune: block.call rune
 
+  /**
+  For every unicode character in the string, call the block.
+  The argument to the block is an integer in the Unicode range of 0-0x10ffff,
+    inclusive.
+  The return value is assembled from the return values of the block.
+  If a string is returned from the block it is inserted at that point in the
+    return value.
+  If a byte array is returned from the block it is converted to a string and
+    treated like a string.  The byte array must contain whole, valid UTF-8
+    sequences.
+  If an integer is returned from the block it is treated as a Unicode code
+    point, and the corresponding code point is inserted.
+  If the block returns null, this is treated like the zero length string.
+  If the block returns a collection, then every element in the collection
+    is handled like the above actions, but this is only done for one level -
+    collections of collections are not flattened in this way.
+  To get a list or byte array as the return value instead of a string, use
+    `str.to_byte_array.map` instead.
+  # Examples.
+  heavy_metalize str/string -> string:
+    return str.map: | c |
+      {'o': 'ö', 'a': 'ä', 'u': 'ü', 'ä': "\u{20db}a"}.get c --if_absent=:c
+  lower_case str/string -> string:
+    return str.map: | c | ('A' <= c <= 'Z') ? c - 'A' + 'a' : c
+  */
+  map [block] -> string:
```

Comment:
Could be called flat_map https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap

Reply:
flat-map would also be an option, but in that case the block would need to return a byte-array. For unicode just returning the rune seems simpler.

Reply:
Sorry. I thought we were still on the ByteArray.

Reply:
Yes. 'flat_map' is definitely better here. Not sure if it should be `map --flatten` or `flat_map`. The latter is more commonly used in literature, but `--flatten` makes the function easier to find.

Reply:
I don't think I want to implement the non-flat version for strings, so the --flatten feels a bit strange.

Reply:
Then `flat_map`.